Sources/X509/Verifier/PolicyBuilder.swift

```diff
@@ -0,0 +1,156 @@
+//
+// This source file is part of the SwiftCertificates open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftCertificates project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftCertificates project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftASN1
+
+@resultBuilder
+public struct PolicyBuilder {
+    @inlinable
+    public static func buildBlock(_ components: some VerifierPolicy) -> some VerifierPolicy {
+        components
+    }
+    
+    @inlinable
+    public static func buildExpression(_ expression: some VerifierPolicy) -> some VerifierPolicy {
+        expression
+    }
+}
+
+
+// MARK: empty policy
+extension PolicyBuilder {
+    @usableFromInline
+    struct Empty: VerifierPolicy {
+        @inlinable
+        var verifyingCriticalExtensions: [SwiftASN1.ASN1ObjectIdentifier] { [] }
+        
+        @inlinable
+        init() {}
+        
+        @inlinable
+        mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain) async -> PolicyEvaluationResult {
+            .meetsPolicy
+        }
+    }
+    
+    @inlinable
+    public static func buildBlock() -> some VerifierPolicy {
+        Empty()
+    }
+}
+
+// MARK: concatenated policies
+extension PolicyBuilder {
+    @usableFromInline
+    struct Tuple2<First: VerifierPolicy, Second: VerifierPolicy>: VerifierPolicy {
+        @usableFromInline
+        var first: First
+        
+        @usableFromInline
+        var second: Second
+        
+        @inlinable
+        init(first: First, second: Second) {
+            self.first = first
+            self.second = second
+        }
+        
+        @inlinable
+        var verifyingCriticalExtensions: [SwiftASN1.ASN1ObjectIdentifier] {
+            first.verifyingCriticalExtensions + second.verifyingCriticalExtensions
+        }
+        
+        @inlinable
+        mutating func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain) async -> PolicyEvaluationResult {
+            switch await first.chainMeetsPolicyRequirements(chain: chain) {
+            case .meetsPolicy:
+                break
+            case .failsToMeetPolicy(let reason):
+                return .failsToMeetPolicy(reason: reason)
+            }
+            
+            return await second.chainMeetsPolicyRequirements(chain: chain)
+        }
+    }
+    
+    @inlinable
+    public static func buildPartialBlock(first: some VerifierPolicy) -> some VerifierPolicy {
+        first
+    }
+    
+    @inlinable
+    public static func buildPartialBlock(accumulated: some VerifierPolicy, next: some VerifierPolicy) -> some VerifierPolicy {
+        Tuple2(first: accumulated, second: next)
+    }
+}
+
+
+// MARK: if
+extension Optional: VerifierPolicy where Wrapped: VerifierPolicy {
```

Comment:
I have slightly mixed feelings about having `Optional` conform to `VerifierPolicy`, I wonder if we should have an internal wrapper type instead.